Re-add ability to gather penalties for all validators

kellybyrd opened the pull request:
This is a re-implementation of 21b8d99e1e28d5e403f65ac267f3b1bf9f4539a8
after the switch to JSON-RPC:

Add an envvar (defaults to off) that will gather penalties for
all validators from the ledger.

Unlike the previous version, this one does not exclude the 0 penalty
validators.

kellybyrd committed:
Re-add ability to gather penalties for all validators

This is a re-implementation of 21b8d99e1e28d5e403f65ac267f3b1bf9f4539a8
after the switch to JSON-RPC:

Add an envvar (defaults to off) that will gather penalties for
all validators from the ledger.

Unlike the previous version, this one does not exclude the 0 penalty
validators.

kellybyrd commented:
@PaulVMo I did this REALLY quick and need to step away for a bit. It appeared to work for at least one case, but I haven't tested it well. I'll do that later.

PaulVMo commented on miner_exporter.py:
```diff
@@ -182,12 +194,13 @@ def stats(miner: MinerJSONRPC):
     if version is not None:
         VALIDATOR_VERSION.labels(my_label).info({'version':version})
 
-    if this_validator is not None:
```
Won't "this validator" also be in the full set? Why don't you apply the same transformations to the miner.ledger_validators(address=addr) result to turn it into a dict of dicts and then you can use one loop to output the ledger details regardless of whether it is a single (this validator) or all validators?

kellybyrd replied:
I don't follow. In the new if/else starting on line 120 above, penalty_ledger contains either either the full set or just "this one". I did that so the for loop below here (starting on 198) can loop over penalty_ledger in all cases. 

The only reason I kept using `this_validator` above was because I saw that other things in the code referenced it and I was moving fast to get this up before I has to leave. `this_validator` ends up being just a copy of one entry from `penalty_ledger`. It's probably cleaner to have everything always refer to `penalty_ledger`

kellybyrd replied:
kk, just fixed this up.

PaulVMo commented on miner_exporter.py:
```diff
@@ -182,12 +194,13 @@ def stats(miner: MinerJSONRPC):
-        LEDGER_PENALTY.labels('ledger_penalties', 'total', my_label).set(this_validator['total_penalty'])
-        BLOCKAGE.labels('last_heartbeat', my_label).set(this_validator['last_heartbeat'])
+    if penalty_ledger is not None:
+        for name, ledger_entry in penalty_ledger.items():
```
Filter out where status != staked?

kellybyrd replied:
Good idea, I'll do this in the dict comprehension right when I collect above.

kellybyrd replied:
done.

PaulVMo commented:
Yeah that's what I was referring to, use penalty_ledger even in the case of
the single validator. I did forget that this_validator was used later

On Mon, Nov 29, 2021, 7:49 PM Kelly Byrd ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In miner_exporter.py
> <https://github.com/PaulVMo/miner_exporter/pull/2#discussion_r758868816>:
>
> > @@ -182,12 +194,13 @@ def stats(miner: MinerJSONRPC):
>      if version is not None:
>          VALIDATOR_VERSION.labels(my_label).info({'version':version})
>
> -    if this_validator is not None:
>
> I don't follow. In the new if/else starting on line 120 above,
> penalty_ledger contains either either the full set or just "this one". I
> did that so the for loop below here (starting on 198) can loop over
> penalty_ledger in all cases.
>
> The only reason I kept using this_validator above was because I saw that
> other things in the code referenced it and I was moving fast to get this up
> before I has to leave. this_validator ends up being just a copy of one
> entry from penalty_ledger. It's probably cleaner to have everything
> always refer to penalty_ledger
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/PaulVMo/miner_exporter/pull/2#discussion_r758868816>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABLJDYNIUZC6VZ5IA7OYAX3UOQUTXANCNFSM5JANVQOQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

kellybyrd committed:
Cleanup from code review

1) Remove use of this_validator completely. This makes things a bit
cleaner.

2) In the ALL_PENALTIES case, filter for only staked validator

PaulVMo committed:
Merge branch 'main' into krb/all_penalties
